SIMD-0105: Maintain Dynamic Set of Reserved Account Keys

jstarry opened the pull request:
The transaction scheduler and the runtime both demote transaction write locks for builtin programs and sysvars using a static list of reserved IDs. This proposal replaces the currently used static lists of builtin program and sysvar IDs with a dynamic set of reserved account keys that can be updated at epoch boundaries with feature gates.

jstarry commented on proposals/0100-read-only-reserved-accounts.md:
```diff
@@ -0,0 +1,149 @@
+    epoch_rewards::id(),
+    last_restart_slot::id(),
+    Pubkey::from_str("ZkTokenProof1111111111111111111111111111111").unwrap(),
+];
```
They're there, I'll replace with base58 keys for clarity

apfitzge commented on proposals/0100-read-only-reserved-accounts.md:
```diff
@@ -0,0 +1,149 @@
+same way as the runtime to avoid scheduling issues. Otherwise if a scheduler
+prematurely demotes transaction write locks before that sysvar exists, it could
+send an invalid batch of conflicting transactions to the runtime.
+
```
Yeah, it seems a dynamic list is the most straight-forward way to get the read-only behavior and keep the scheduler aware for optimization.

t-nelson replied:
fwiw, i'm in support of failing txes that writelock reserved addresses at accounts load time (in a different simd). it was cowardly that we didn't do this in the first place. the cluster ought not burden itself with resolving hanlon's razor. it's a negative net effect whether done by incompetence or malice

jstarry replied:
I'm also in support and agree we don't need to hand hold this much. I will think about it for a future SIMD but for now I have concluded we need a dynamic set of keys even if we decide to start failing those transactions.

jacobcreech commented on proposals/0100-read-only-reserved-accounts.md:
```diff
@@ -0,0 +1,140 @@
+---
+simd: "0100"
```
```suggestion
simd: "0105"
```

apfitzge replied:
PR title should also be udpated

CriesofCarrots commented on proposals/0100-read-only-reserved-accounts.md:
```diff
@@ -0,0 +1,140 @@
+- BPFLoader2111111111111111111111111111111111
+- BPFLoaderUpgradeab1e11111111111111111111111
+
+The set of reserved keys may only be modified via feature gate activation. On
+epoch boundaries, validator implementations should add or remove reserved keys
+as dictated by feature gated code. The first feature gate activated modification
```
How will/should accounts like those added in https://github.com/solana-labs/solana/pull/34624 be included? Presumably they ought to have write-locks demoted, but there will be a new one every epoch, which probably isn't feasible to feature-gate.

jstarry replied:
I think it's ok if those accounts are able to be write locked for now.

t-nelson replied:
agreed, it's probably fine to take writelocks on these accounts. runtime will be doing its writing before transaction processing begins so there will be no contention there. 

---

the only reason i can think of to reference these accounts on chain would be to determine if an epoch rewards distribution is ongoing at execution time. this would require the ability to query block height, which we don't have. should that be made available in the future, we could add only the current epoch's epoch rewards partition metadata account address to the reserved set.

jstarry replied:
We already have the epoch rewards sysvar available for checking if rewards distribution is happening so that case is covered.

t-nelson replied:
oh i thought we killed that. kinda... not necessary for anything other than `is_in_epoch_rewards_period()`, for which it's overkill

jstarry replied:
@CriesofCarrots told me that the sysvar is sticking around

CriesofCarrots replied:
Fwiw, I can imagine a scenario in which an on-chain program would want to get the partitioned rewards data from one of these accounts immediately on the epoch boundary in order to do something as soon as a reward is delivered, instead of waiting until the end of the distribution period. But it's not a super compelling use-case, and we can certainly enforce that such programs wait.

That said, my question was meant to be a general one. These new PDA sysvars aren't fundamentally different than statically addressed sysvars, and there may be others in the future that have a stronger on-chain use-case. Just kick the can?

jstarry replied:
> [...] in order to do something as soon as a reward is delivered, instead of waiting until the end of the distribution period

Stake accounts can't be modified until the end of the distribution period so I don't think there's much value in doing anything immediately at the epoch boundary.

> These new PDA sysvars [...]

The PDAs you created for epoch rewards are stake program owned, aren't they? The stake program PDAs aren't much different from things like the Jito validator history program which creates a bunch of accounts to store historical validator performance. I don't think the epoch rewards PDA need to be enshrined as read-only at this point...

> Just kick the can?

Yup! We can address further write lock controls in a new SIMD for these types of use cases. This SIMD is for a few commonly used accounts that could disrupt a large chunk of transactions if they could be write locked. The approach relies on an in-memory set of accounts which is not designed to grow a lot.

CriesofCarrots replied:
> Stake accounts can't be modified until the end of the distribution period so I don't think there's much value in doing anything immediately at the epoch boundary.

I'm a bit surprised you can look at the state of high-speed trading/speculation and make that argument. We're talking over a minute here.

> We can address further write lock controls in a new SIMD for these types of use cases.

Okay, let's do that :)

t-nelson replied:
> look at the state of high-speed trading/speculation

omg there are going to be options on newly distributed validator rewards :handshake:

t-nelson commented on proposals/0100-read-only-reserved-accounts.md:
```diff
@@ -0,0 +1,140 @@
+
+The main consideration is making sure that all `is_writable` checks are
+consistent before and after implementing this proposal to avoid breaking
+consensus.
```
i'd add a note about this not scaling to too many reserved addresses. should we decide that we need to provide this behavior to a large number of accounts in the future, we should probably actually flip a bit in the accounts' metadata

jstarry replied:
Agreed. In the future I would like the account metadata to designate a write lock authority.

t-nelson reviewed (approved):
just one nit from me (security section), otherwise just commentary. thanks!

jstarry commented:
I addressed @t-nelson's feedback and updated the proposal file name. Any other comments? @CriesofCarrots do you think it's reasonable to allow write locks for now on epoch rewards PDA accounts?

steviez reviewed (approved):
Re-adding my approval; I agree with Justin's latest comments about further write-lock controls for epoch rewards PDA accounts (and/or similar cases) being considered in a separate SIMD.

While the motivation is similar (and maybe the implementation too), I think the cases are distinct enough to justify handling them separately. Plus, handling them separately will keep each change smaller and easier to reason about.

That all being said, I'd like to get explicit confirmation from CriesofCarrots that she is onboard with this plan before proceeding. I'm not as familiar with the epoch rewards piece, so there may be subtlety that I'm missing.

ptaffet-jump commented:
FYI, our current plan for Firedancer is to refuse to include transactions that would trigger the writelock demotion in blocks we build. We'll still be able to replay them, so it won't break consensus, but adding the demotion seems like additional complexity for no good reason. I'd definitely be in favor of the proposal Trent alluded to earlier of making these transactions fail.
